### PR TITLE
Pop tail on cancel

### PR DESCRIPTION
**Pull request by kelvinlau20100**
Add ability to pop events from the event q but only permissioned

### CODE REVIEW

File: dex/src/state.rs [hunk 1]
```diff
@@ -3411,6 +3497,98 @@ impl State {
         Ok(())
     }
```
Comment (kelvinlau20100) [inline dex/src/state.rs:3499]: I think one thing that could be done here is to move this repeated logic where you mutate the open orders account to a separate function but also I think fine as is

Comment (kelvinlau20100) [reply dex/src/state.rs:3516]: fixed this

File: dex/src/state.rs [hunk 2]
```diff
@@ -887,6 +887,27 @@ impl<'a, H: QueueHeader> Queue<'a, H> {
...
+    #[inline]
+    pub fn peek_back(&self) -> Option<&H::Item> {
+        if self.empty() {
+            return None;
+        }
+        Some(&self.buf[(self.header.head() + self.header.count() - 1) as usize])
+    }
+
+    #[inline]
+    pub fn pop_back(&mut self) -> Result<H::Item, ()> {
```
Comment (konrads) [inline dex/src/state.rs:899]: shouldn't this be an option too?

Comment (kelvinlau20100) [reply dex/src/state.rs:899]: I think it should error in the case where there is no event to pop

File: dex/src/state.rs [hunk 3]
```diff
@@ -2225,6 +2246,63 @@ pub(crate) mod account_parser {
...
+        pub event_q: EventQueue<'a>,
+    }
+    impl<'a, 'b: 'a> PopTailOutEventsArgs<'a, 'b> {
+        pub fn with_parsed_args_permissioned<T>(
+            program_id: &'a Pubkey,
+            accounts: &'a [AccountInfo<'b>],
+            limit: u16,
+            f: impl FnOnce(PopTailOutEventsArgs) -> DexResult<T>,
+        ) -> DexResult<T> {
+            check_assert!(accounts.len() >= 4)?;
```
Comment (xjcaa) [inline dex/src/state.rs:2265]: 5?

File: dex/src/state.rs [hunk 4]
```diff
@@ -3411,6 +3514,79 @@ impl State {
...
+                    native_qty_paid: _,
+                    native_qty_received: _,
+                    native_fee_or_rebate: _,
+                    fee_tier: _,
+                    order_id: _,
+                    owner: _,
+                    owner_slot: _,
+                    client_order_id: _,
+                } => {
+                    return Err(DexErrorCode::InvalidEventToPopFromTail.into());
```
Comment (xjcaa) [inline dex/src/state.rs:3558]: So you only pop from tail when there's an out in Zeta right?

**Review by xjcaa — APPROVED** [review]

**Pull request merged**